cristin-import/src/main/java/no/unit/nva/cristin/lambda/CristinEntryEventConsumer.java

```diff
@@ -54,117 +56,153 @@ public class CristinEntryEventConsumer extends EventHandler<EventReference, Publ
         super(EventReference.class);
         this.resourceService = resourceService;
         this.s3Client = s3Client;
     }
-    
+
     @Override
     protected Publication processInput(EventReference input,
                                        AwsEventBridgeEvent<EventReference> event,
                                        Context context) {
-        
+
         validateEvent(event);
         var eventBody = readEventBody(input);
         return attempt(() -> parseCristinObject(eventBody))
                    .map(CristinObject::toPublication)
                    .flatMap(this::persistInDatabase)
+                   .map(publication -> storePublicationIdentifierAndCristinIdInS3(publication, eventBody))
                    .orElseThrow(fail -> handleSavingError(fail, eventBody));
     }
-    
-    private FileContentsEvent<JsonNode> readEventBody(EventReference input) {
-        var s3Driver = new S3Driver(s3Client, input.extractBucketName());
-        var json = s3Driver.readEvent(input.getUri());
-        return FileContentsEvent.fromJson(json, JsonNode.class);
-    }
-    
+
     @JacocoGenerated
     private static S3Client defaultS3Client() {
         return S3Client.builder()
                    .httpClient(UrlConnectionHttpClient.create())
                    .build();
     }
-    
+
     @JacocoGenerated
     private static AmazonDynamoDB defaultDynamoDbClient() {
         return AmazonDynamoDBClientBuilder
                    .standard()
                    .withRegion(ApplicationConstants.AWS_REGION.id())
                    .build();
     }
-    
+
+    private static Optional<String> getCristinIdentifier(Publication publication) {
+        return publication.getAdditionalIdentifiers()
+                   .stream()
+                   .filter(
+                       additionalIdentifier -> isCristinAdditionalIdentifier(additionalIdentifier))
+                   .findFirst()
+                   .map(additionalIdentifier -> additionalIdentifier.getValue());
+    }
+
+    private static boolean isCristinAdditionalIdentifier(AdditionalIdentifier additionalIdentifier) {
+        return CristinObject.IDENTIFIER_ORIGIN.equals(additionalIdentifier.getSource());
+    }
+
+    private Publication storePublicationIdentifierAndCristinIdInS3(Publication publication,
+                                                                   FileContentsEvent<JsonNode> eventBody) {
+        var cristinIdentifier =
+            getCristinIdentifier(publication);
+        var fileUri = constructResourcehandleFileUri(eventBody, publication);
+        var s3Driver = new S3Driver(s3Client, fileUri.getUri().getHost());
+        attempt(() -> s3Driver.insertFile(fileUri.toS3bucketPath(),
+                                          cristinIdentifier.orElse(StringUtils.EMPTY_STRING))).orElseThrow();
+        return publication;
+    }
+
+    private UriWrapper constructResourcehandleFileUri(FileContentsEvent<JsonNode> eventBody, Publication publication) {
```

Comment:
Resourcehandle? Not sure what this means

Reply:
Ups. copy-paste from brage-import. Does "constructSuccessFileUri" sound better? It's more inline with the almost identical function "constructErrorFileUri".